lib/travis/remote_log.rb

```diff
@@ -0,0 +1,350 @@
+
+    attribute :aggregated_at, Time
+    attribute :archive_verified, Boolean, default: false
+    attribute :archived_at, Time
+    attribute :archiving, Boolean, default: false
+    attribute :content, String
+    attribute :created_at, Time
+    attribute :id, Integer
+    attribute :job_id, Integer
+    attribute :purged_at, Time
+    attribute :removed_at, Time
+    attribute :removed_by_id, Integer
+    attribute :updated_at, Time
+
+    def platform=(platform)
+      @platform = platform
+    end
+
+    def platform
+      @platform || :default
+    end
+
+    def job
+      @job ||= Job.find(job_id)
+    end
+
+    def removed_by
+      return nil unless removed_by_id
+      @removed_by ||= User.find(removed_by_id)
+    end
+
+    def removed?
+      !removed_by_id.nil?
+    end
+
+    def parts(after: nil, part_numbers: [])
+      return solo_part if removed? || aggregated?
+      remote.find_parts_by_job_id(
+        job_id, after: after, part_numbers: part_numbers
+      )
+    end
+
+    alias log_parts parts
+
+    private def solo_part
+      [
+        RemoteLogPart.new(
+          number: 0,
+          content: content,
+          final: true
+        )
+      ]
+    end
+
+    def aggregated?
+      !!aggregated_at
+    end
+
+    def archived?
+      !!(!archived_at.nil? && archive_verified?)
+    end
+
+    def archived_url(expires: nil)
+      @archived_url ||= remote.fetch_archived_url(job_id, expires: expires)
+    end
+
+    def archived_log_content
+      @archived_content ||= remote.fetch_archived_log_content(job_id)
+    end
+
+    def to_json(chunked: false, after: nil, part_numbers: [])
+      as_json(
+        chunked: chunked,
+        after: after,
+        part_numbers: part_numbers
+      ).to_json
+    end
+
+    def as_json(chunked: false, after: nil, part_numbers: [])
+      ret = {
+        'id' => id,
+        'job_id' => job_id,
+        'type' => 'Log'
+      }
+
+      unless removed_at.nil?
+        ret['removed_at'] = removed_at.utc.to_s
+        ret['removed_by'] = removed_by.name || removed_by.login
+      end
+
+      if chunked
+        ret['parts'] = parts(
+          after: after,
+          part_numbers: part_numbers
+        ).map(&:as_json)
+      else
+        ret['body'] = archived_log_content || content
+      end
+
+      { 'log' => ret }
+    end
+
+    def clear!(user = nil)
+      message = ''
+      removed_by = nil
+
+      if user.respond_to?(:name) && user.respond_to?(:id)
+        message = "Log removed by #{user.name} at #{Time.now.utc}"
+        removed_by = user.id
+      end
+
+      updated = remote.write_content_for_job_id(
+        job_id,
+        content: message,
+        removed_by: removed_by
+      )
+
+      attributes.keys.each do |k|
+        send("#{k}=", updated.send(k))
+      end
+
+      message
+    end
+
+    def remote
+      @remote ||= Travis::RemoteLog::Remote.new(platform: platform)
+    end
+
+    class Client
```

Comment:
One of the task requirements was that in case the log is archived in the S3 bucket (the content is not present in the database) we should make a specific call using the remote log from travis-api to the S3 bucket and retrieve the archived log content.

Reply:
@Jullianos @AndriiMysko @speedywizard if you already using S3 client lib to upload files directly to AWS S3, why you are not using the same S3 client lib to get the files?

Reply:
In regards to this I've already cleared the remote_log.rb file from all the unnecessary code with my latest commit.